app/api/callback/route.ts

```diff
@@ -0,0 +1,57 @@
+import { saveSession } from 'app/auth';
+
+export async function GET(request: Request) {
+  let { searchParams } = new URL(request.url);
+  let code = searchParams.get('code');
+  let state = searchParams.get('state');
+
+  if (!code || !state) {
```

Comment:
The `state` value needs to be compared against the value that was sent in the `authorization` request.

One way of doing it would be to set an cookie `HttpOnly` cookie, do the comparison here, and then delete the cookie after to clean up.